File: src/oid.ts
/**
 * oid.ts -- OID computation for GAP CDROs.
 *
 * The CDRO OID content-core projection is IMPORTED from @synoi/sraid, not
 * reimplemented here.
 *
 * It used to be reimplemented, derived faithfully from
 * synoi-sraid/PROJECTION_SPEC.md, with the correct six-name strip set, on the
 * stated assumption that "a byte-for-byte divergence from the reference
 * implementation is caught by the shared conformance vectors". That assumption
 * was false in both halves. The copy diverged - it carried the same __proto__
 * OID collision as the reference, because the prose "remove exactly these six
 * fields, keep everything else" translates in JavaScript into `core[k] = v`,
 * which silently drops an own __proto__ - and no conformance vector covered
 * that shape, so nothing caught it. Two teams reading correct prose wrote the
 * same defect.
 *
 * The original reason for the copy was real: sraid's default entry imports
 * node:crypto and this package must stay portable. The projection never needed
 * crypto - only the hashing did - so it now comes from the PURE
 * @synoi/sraid/content-core subpath while the hash stays @noble here. Identical
 * by construction rather than by agreement.
 *
 *     OID = "sha256:" + hex(sha256(canonicalize(cdroContentCore(object))))
 *
 * cdroContentCore removes EXACTLY the six detached-signature / envelope fields
 * (see CDRO_ENVELOPE_FIELDS) at the top level and KEEPS everything else -- in
 * particular gap_version and supersedes are KEPT and hashed into identity, so
 * a protocol downgrade or a lineage-edge tamper changes the OID. This is the
 * one projection that yields the SAME OID whether the object is pre- or
 * post-attestation, which is what lets a third party recompute the OID of a
 * SIGNED receipt and match the value the signer stamped.
 */

import { sha256 } from '@noble/hashes/sha256'
import { canonicalize } from './canonicalize.js'
// THE normative projection, imported rather than reimplemented. Pure subpath:
// no node:crypto in its graph, so this package stays portable and keeps its
// own @noble hash. See the header note above.
import { cdroContentCore } from '@synoi/sraid/content-core'

/** Convert a Uint8Array to a lowercase hex string. */
function bytesToHex(bytes: Uint8Array): string {
  const hex: string[] = []
  for (let i = 0; i < bytes.length; i++) {
    const b = bytes[i] as number
    hex.push((b >>> 4).toString(16))
    hex.push((b & 0x0f).toString(16))
  }
  return hex.join('')
}

/**
 * The detached-signature / envelope fields removed by `cdroContentCore` before
 * hashing. This is the ONE normative strip-set for the CDRO OID projection
 * (ADR_019 decision 1; PROJECTION_SPEC.md §2). It is derived from the reference
 * implementation @synoi/sraid `CDRO_ENVELOPE_FIELDS`, NOT independently
 * hand-listed.
 *
 * The set is SEMANTIC: "every field produced BY the signer after
 * canonicalization, plus the OID output itself." Concretely:
 *
 *   oid                  - the projection OUTPUT (cannot be an input to itself)
 *   signature            - legacy hybrid SignatureEnvelope (attaches after hash)
 *   ml_dsa_signature     - detached PQ signature (attaches after hash)
 *   signature_key_id     - signer-stamped key id (produced by the signer)
 *   signature_algorithm  - signer-stamped alg id (produced by the signer)
 *   attestation          - DSSE AttestationEnvelope (attaches after hash)
 *
 * EVERYTHING ELSE IS KEPT and hashed into identity, including in particular
 * `gap_version` (so a protocol downgrade is OID-detectable) and `supersedes`
 * (so the SRAID Merkle-DAG lineage edge is tamper-evident).
 */
/**
 * Compute the OID of a GAP CDRO.
 *
 * Accepts either a pre-stripped payload or a full envelope. The six detached
 * envelope fields (CDRO_ENVELOPE_FIELDS) are stripped before canonicalization
 * so both forms produce the same OID and a post-attestation object yields the
 * SAME OID as its pre-attestation form.
 *
 * @param body - the OID payload or full CDRO envelope.
 * @returns the canonical OID string `"sha256:<hex>"`.
 */
export function computeGapOid(body: unknown): string {
  const canonical = canonicalize(cdroContentCore(body))
  const digest = sha256(new TextEncoder().encode(canonical))
  return 'sha256:' + bytesToHex(digest)
}

File: rust/src/oid.rs
//! OID computation and CDRO content-core projection.
//!
//! OID = "sha256:" + lowercase_hex(sha256(canonicalize(content))).
//!
//! This is DERIVED from the single normative source,
//! synoi-sraid/PROJECTION_SPEC.md (ADR_019 decisions 1-3), and matches the
//! reference implementation byte-for-byte:
//!   - `@synoi/sraid` `oidOf` / `cdroContentCore` (src/oid.ts)
//!   - `@synoi/gap`   `computeGapOid`             (src/oid.ts)
//!
//! There is exactly ONE content-core projection (`cdro_content_core`): it
//! strips the SIX detached-signature / envelope fields
//! {oid, signature, ml_dsa_signature, signature_key_id, signature_algorithm,
//! attestation} and KEEPS everything else, including `gap_version` (protocol
//! downgrade is OID-detectable) and `supersedes` (Merkle-DAG lineage edge is
//! tamper-evident). `compute_gap_oid` computes this ONE projection (same result
//! as `cdro_oid` for object inputs) and is kept under its historical name for
//! API stability; both yield the same OID a third party recomputes for a signed
//! (post-attestation) receipt. Both enforce the GAP number rule (floats
//! rejected). Used by `gap/oid.json` and `sraid/cdro-roundtrip.json`.
//!
//! The bare `sraid/oid.json` vectors hash the input value directly with no field
//! stripping (`oidOf`), which is `oid_of` here.

use serde_json::Value;
use sha2::{Digest, Sha256};

use crate::canonicalize::{canonicalize, canonicalize_gap};

/// Compute `sha256:<hex>` over the canonical form of a value, with no field
/// stripping. Counterpart of `@synoi/sraid` `oidOf`.
pub fn oid_of(value: &Value) -> String {
    oid_of_canonical(&canonicalize(value))
}

/// Compute `sha256:<hex>` over already-canonicalized bytes. Counterpart of
/// `@synoi/sraid` `oidOfCanonical`.
pub fn oid_of_canonical(canonical: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(canonical.as_bytes());
    let digest = hasher.finalize();
    format!("sha256:{}", hex::encode(digest))
}

/// The SIX detached-signature / envelope fields removed by the content-core
/// projection before hashing (ADR_019 decision 1; PROJECTION_SPEC.md §2).
/// gap_version and supersedes are deliberately NOT here -- they are KEPT in
/// identity. Mirrors `@synoi/sraid` `CDRO_ENVELOPE_FIELDS`.
const CDRO_ENVELOPE_FIELDS: &[&str] = &[
    "oid",
    "signature",
    "ml_dsa_signature",
    "signature_key_id",
    "signature_algorithm",
    "attestation",
];

/// Project a CDRO (or a pre-stripped OID payload) to its content core by
/// removing the six detached fields at the TOP LEVEL only. Non-object inputs
/// are returned unchanged, exactly as the TS reference does.
pub fn cdro_content_core_value(value: &Value) -> Value {
    strip_top_level(value, CDRO_ENVELOPE_FIELDS)
}

/// Compute the GAP CDRO OID: strip the six detached fields, then hash the
/// float-rejecting canonical form. Counterpart of `@synoi/gap` `computeGapOid`
/// / `@synoi/sraid` `cdroOid`. Kept under the historical name for API
/// stability; it computes the ONE normative content-core projection and
/// enforces the GAP number rule (ADR_019 decision 2).
///
/// A non-integer number is out-of-domain for a GAP CDRO (money is integer
/// minor units, time is integer milliseconds); it is rejected loud, matching
/// the TS/Python SDKs which throw. Use [`try_compute_gap_oid`] for a fallible
/// variant that returns the typed error instead of panicking.
pub fn compute_gap_oid(value: &Value) -> String {
    try_compute_gap_oid(value).unwrap_or_else(|e| panic!("compute_gap_oid: {e}"))
}

/// Fallible form of [`compute_gap_oid`]: returns the typed float-reject error
/// instead of panicking. Mirrors the GAP number rule (ADR_019 §3.1).
pub fn try_compute_gap_oid(value: &Value) -> Result<String, String> {
    let core = cdro_content_core_value(value);
    Ok(oid_of_canonical(&canonicalize_gap(&core)?))
}

/// Project a full CDRO to its content core by removing the six detached
/// fields at the top level. The argument MUST be a JSON object; mirrors the
/// `TypeError` the TS reference throws otherwise.
pub fn cdro_content_core(value: &Value) -> Result<Value, String> {
    if !value.is_object() {
        return Err("cdro_content_core: argument must be a CDRO object".to_string());
    }
    Ok(cdro_content_core_value(value))
}

/// Compute the OID of a full CDRO over its content core. Counterpart of
/// `@synoi/sraid` `cdroOid`. Identical result to `compute_gap_oid` for object
/// inputs; this variant errors on non-object input AND on any non-integer
/// number (ADR_019 §3.1), returning the typed error rather than panicking.
pub fn cdro_oid(value: &Value) -> Result<String, String> {
    let core = cdro_content_core(value)?;
    Ok(oid_of_canonical(&canonicalize_gap(&core)?))
}

fn strip_top_level(value: &Value, excluded: &[&str]) -> Value {
    match value {
        Value::Object(map) => {
            let mut out = serde_json::Map::new();
            for (k, v) in map {
                if !excluded.contains(&k.as_str()) {
                    out.insert(k.clone(), v.clone());
                }
            }
            Value::Object(out)
        }
        other => other.clone(),
    }
}
